### DISCUSSION TITLE
Proposal: generic constraint allows static classes

### DISCUSSION THREAD

**Discussion by sgjsakura** [dotnet/csharplang#4840] [Language Ideas]
## Background

Currently static classes are not allowed to be used as generic type arguments. This limitation may be introduced since you can use the type argument as variable or parameter types, while them with static classes are invalid. 

However, sometimes library providers uses generic types just as a simplification for passing `Systesm.Type` object references and thus the it will be used only for the `typeof` expression. Under such circumstance, the general limitation for type arguments can be removed for better type comaptibility. 

An existing example is the `IStringLocalizer<T>` class which is provieded by the ASP.NET Core Localization official library. The type parameter `T` is used for locating the managed resource file with the same name. The type `T` will be used only under the `typeof` expression, and variable of type T will never be referenced. Actually, you can use the low level provider `IStringLocalizerFactory` and passing `typeof(MyStaticClass)` as the argument type, while the current syntax limitation does not support to simplify your code with generic arguments.

## Suggested Syntax

The suggested syntax for declaring a type parameter "constraint" that allows static classes maybe one of the following:
```C#
public class MyType<T> where T : static
{
}
```
or 
```C#
public class MyType<T> where T : any
```
We can either choose to reuse the triditional keyword `static`, or introduce a new keyword `any` which may be better for code semantics.

## Compatibility and Limitation

Obviously, the static generic constraint will introduce extra limitation for the argument usage. More specifically, this type argument can only be used inside the `typeof` expression, while any following usage should be forbidden:
1. Declare any parameter or variable with this type;
2. Combine this constraint with any other ones (`class`, `struct`, `not null` and so on)
3. Use this parameter as generic arguments  without the same constraint of other types.

Relatively, receiving generic type arguments from other ones with any kind of constraints should be allowed since the static constraint actually has the minimum usage requirement.

**Comment by huoyaoyuan** [comment]
Generally, this should be even less useful to allow `ref struct` or `void` being generic parameter. Another use case I've found was `System.Windows.WeakEventManager`.

We are now having static virtual methods. However, static class can't inherit or implement interface.

**Reply by JustNrik** [reply]
`void` being a generic parameters makes a lot of sense so you shouldn't put in it in the same place as `ref struct`.

**Comment by Unknown6656** [comment]
This would only make sense in combination with static interfaces (or at least static virtual members), otherwise how would you effectively use `T` inside your method (apart from `typeof(T)`)?

**Comment by Richiban** [comment]
Rather than this, wouldn't it be better to support top level object declarations the way Kotlin and Scala can?

An object declaration is something between a singleton type and a static class: it looks and feels like a static class when you write it but you can still use it like a real type: you can pass references to it and put it in a list but it's a singleton and there's only one value of this type (and exactly one value).

Maybe static classes could be retconned to mean something like this?

**Comment by jnm2** [comment]
This proposal is included under option 2 in Stephen Toub's proposal at https://github.com/dotnet/csharplang/issues/5783.
